services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/UploadPartRequestPublisher.java

```diff
@@ -0,0 +1,379 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.services.s3.internal.multipart;
+
+import java.nio.ByteBuffer;
+import java.util.Optional;
+import java.util.concurrent.BlockingDeque;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.core.async.AsyncRequestBody;
+import software.amazon.awssdk.core.async.SdkPublisher;
+import software.amazon.awssdk.core.exception.SdkClientException;
+import software.amazon.awssdk.core.internal.util.NoopSubscription;
+import software.amazon.awssdk.services.s3.internal.crt.RequestConversionUtils;
+import software.amazon.awssdk.services.s3.model.PutObjectRequest;
+import software.amazon.awssdk.services.s3.model.UploadPartRequest;
+import software.amazon.awssdk.utils.Logger;
+import software.amazon.awssdk.utils.NumericUtils;
+import software.amazon.awssdk.utils.Pair;
+import software.amazon.awssdk.utils.async.ByteBufferStoringSubscriber;
+import software.amazon.awssdk.utils.async.SimplePublisher;
+
+/**
+ * Generates {@link UploadPartRequest} and the associated {@link AsyncRequestBody} pairs.
+ * Upon subscription request, it will start to prepare requests asynchronously by loading content from the provided
+ * {@link AsyncRequestBody}. The maximum memory consumed is {@code numOfPartsBuffered * partSize} MB.
+ */
+@SdkInternalApi
+public final class UploadPartRequestPublisher implements SdkPublisher<Pair<UploadPartRequest, AsyncRequestBody>> {
```

Comment:
Can you please comment on all of the methods what they're supposed to be doing? It's tough to keep this all in my head when I'm working from just code.